homeassistant/components/elmax/__init__.py

```diff
@@ -0,0 +1,284 @@
+        return self._panel_entry
+
+    @property
+    def panel_status(self) -> PanelStatus | None:
+        """Return the last fetched panel status."""
+        return self.data
+
+    def get_endpoint_state(self, endpoint_id: str) -> DeviceEndpoint | None:
+        """Return the last fetched status for the given endpoint-id."""
+        if self._state_by_endpoint is not None:
+            return self._state_by_endpoint.get(endpoint_id)
+        return None
+
+    @property
+    def http_client(self):
+        """Return the current http client being used by this instance."""
+        return self._client
+
+    async def _async_update_data(self):
+        try:
+            async with async_timeout.timeout(DEFAULT_TIMEOUT):
+                # Retrieve the panel online status first
+                panels = await self._client.list_control_panels()
+                panels = list(filter(lambda x: x.hash == self._panel_id, panels))
+
+                # If the panel is no more available within the given. Raise config error as the user must
+                # reconfigure it in order to  make it work again
+                if len(panels) < 1:
+                    _LOGGER.error(
+                        "Panel ID %s is no more linked to this user account",
+                        self._panel_id,
+                    )
+                    raise ConfigEntryAuthFailed()
+
+                panel = panels[0]
+                self._panel_entry = panel
+
+                # If the panel is online, proceed with fetching its state
+                # and return it right away
+                if panel.online:
+                    status = await self._client.get_panel_status(
+                        control_panel_id=panel.hash, pin=self._panel_pin
+                    )  # type: PanelStatus
+
+                    # Store a dictionary for fast endpoint state access
+                    self._state_by_endpoint = {
+                        k.endpoint_id: k for k in status.all_endpoints
+                    }
+                    return status
+
+                # Otherwise, return None. Listeners will know that this means the device is offline
+                return None
+
+        except ElmaxBadPinError as err:
+            _LOGGER.error("Control panel pin was refused")
+            raise ConfigEntryAuthFailed from err
+        except ElmaxBadLoginError as err:
+            _LOGGER.error("Refused username/password")
+            raise ConfigEntryAuthFailed from err
+        except ElmaxApiError as err:
+            raise HomeAssistantError(f"Error communicating with ELMAX API: {err}")
+        except ElmaxNetworkError as err:
+            raise HomeAssistantError(
+                "Network error occurred while contacting ELMAX cloud"
+            ) from err
+        except Exception as e:
+            _LOGGER.exception("Unexpected exception")
+            raise HomeAssistantError("An unexpected error occurred") from e
+
+
+class ElmaxEntity(Entity):
```

Comment:
I think it is clearer if you move this class to a separate file (eg. `common.py` or `elmax_entity.py`)

Reply:
Agreed!